File: v1/examples/python/09_push.py
"""
09_push.py - TS-API v1 Push Event

Endpoint:
  POST /api/v1/push
      - Push external events into the NVR
      - Requires Push license enabled on the NVR

Event Types:
  1. LPR - Push a license plate recognition result
  2. emergencyCall - Trigger emergency call alarm

  WARNING: emergencyCall with event=callStart triggers a REAL alarm on the NVR.
           The alarm persists until callEnd is sent. Always send callEnd after callStart.

  IMPORTANT: For emergencyCall, you MUST send callEnd to stop the alarm.
             Forgetting callEnd leaves the NVR in alarm state.
"""

import os
import sys
import time

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config import NVR_API_KEY
from http_client import NvrClient

if not NVR_API_KEY:
    print('NVR_API_KEY environment variable is required')
    sys.exit(1)


def push_lpr_event(client):
    """Push an LPR (license plate recognition) event."""
    print("=== Push LPR Event ===")
    payload = {
        'topic': 'LPR',
        'plateNo': '12AB3456',
        'src': 'booth-01',          # Source booth code (string)
        'when': '2026-01-15 10:30:00',
    }
    r = client.post('/api/v1/push', json=payload)
    print(f"  Status: {r.status_code}")
    if r.status_code == 200:
        print(f"  Response: {r.json()}")
    else:
        print(f"  Ensure Push license is enabled")


def push_emergency_call(client):
    """Push an emergency call event.

    WARNING: callStart triggers a REAL alarm on the NVR!
    You MUST send callEnd to stop it.
    """
    print("\n=== Push Emergency Call ===")
    print("  WARNING: This triggers a REAL alarm on the NVR!")

    # Uncomment the lines below to actually trigger the alarm.
    # Make sure you always send callEnd after callStart.

    # --- Start the emergency call ---
    # print("  Sending callStart...")
    # r = client.post('/api/v1/push', json={
    #     'topic': 'emergencyCall',
    #     'event': 'callStart',
    #     'device': 'intercom-01',   # Device identifier
    #     'src': 'lobby-entrance',   # Source identifier
    # })
    # print(f"  callStart -> {r.status_code}")
    #
    # time.sleep(3)  # Alarm is active for 3 seconds
    #
    # --- MUST send callEnd to stop the alarm ---
    # print("  Sending callEnd...")
    # r = client.post('/api/v1/push', json={
    #     'topic': 'emergencyCall',
    #     'event': 'callEnd',
    #     'device': 'intercom-01',   # Device identifier
    #     'src': 'lobby-entrance',   # Source identifier
    # })
    # print(f"  callEnd -> {r.status_code}")

    print("  (Commented out for safety - uncomment to test)")
    print("  IMPORTANT: Always send callEnd after callStart!")


def main():
    client = NvrClient()
    client.set_api_key(NVR_API_KEY)

    push_lpr_event(client)
    push_emergency_call(client)


if __name__ == '__main__':
    main()

File: v1/examples/csharp/09_Push.cs
// 09_Push.cs - v1 API Push Notification Example
//
// Send events from an external system to the NVR.
//
// Endpoints:
//   POST /api/v1/push  - Send push event
//
// Supported topics:
//   - "LPR": License plate recognition event
//   - "emergencyCall": Emergency call event
//
// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
// !! WARNING:                                                       !!
// !! emergencyCall with callStart triggers an ACTUAL alarm!         !!
// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
//
// Prerequisite: Push license must be enabled.

using System;
using System.Threading.Tasks;

namespace TsApiExamples.V1
{
    class Push
    {
        static async Task Main(string[] args)
        {
            using var client = new NvrClient();
            if (string.IsNullOrEmpty(client.ApiKey))
            {
                Console.WriteLine("NVR_API_KEY environment variable is required");
                return;
            }
            client.SetApiKey(client.ApiKey);

            // ── Send LPR event ──
            Console.WriteLine("=== Sending LPR Push Event ===");

            var lprEvent = new
            {
                topic = "LPR",
                src = "ExternalCam1",           // Source name
                plateNo = "12A3456",            // Plate number
                when = DateTime.Now.ToString("yyyyMMddHHmmss")
            };

            var lr = await client.PostJsonAsync("/api/v1/push", lprEvent);
            Console.WriteLine($"  Status: {(int)lr.StatusCode} {lr.StatusCode}");
            var lBody = await NvrClient.ReadStringAsync(lr);
            if (!string.IsNullOrEmpty(lBody))
                Console.WriteLine($"  Response: {lBody}");

            // ── Emergency call event ──
            Console.WriteLine("\n=== Emergency Call Event ===");
            Console.WriteLine("  WARNING: callStart triggers a real alarm!");

            // callStart example - triggers real alarm!
            var callStart = new
            {
                topic = "emergencyCall",
                device = "intercom-01",
                src = "Lobby",
                @event = "callStart",
                camera = 1
            };
            Console.WriteLine("  (callStart is commented out)");
            // Uncomment to trigger actual alarm!
            // var csr = await client.PostJsonAsync("/api/v1/push", callStart);
            // Console.WriteLine($"  callStart status: {(int)csr.StatusCode}");

            // callEnd example - end alarm
            var callEnd = new
            {
                topic = "emergencyCall",
                device = "intercom-01",
                src = "Lobby",
                @event = "callEnd",
                camera = 1
            };

            Console.WriteLine("  Sending callEnd...");
            var cer = await client.PostJsonAsync("/api/v1/push", callEnd);
            Console.WriteLine($"  Status: {(int)cer.StatusCode} {cer.StatusCode}");
        }
    }
}
